File: src/common/Helpers/HelpHelpers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class HelpHelpers
{
    public static IEnumerable<string> GetHelpTopics()
    {
        var allResourceNames = EmbeddedFileHelpers.GetEmbeddedStreamFileNames();

        var helpPrefix = $"{ProgramInfo.Name}.help.";
        var helpTopics = allResourceNames
            .Where(name => name.StartsWith(helpPrefix, StringComparison.OrdinalIgnoreCase))
            .Select(name => name.Substring(helpPrefix.Length))
            .Select(name => name.Substring(0, name.Length - ".txt".Length))
            .Distinct()
            .OrderBy(name => name.Count(x => x == ' ').ToString("000") + name)
            .ToList();

        return helpTopics;
    }

    public static bool HelpTopicExists(string topic)
    {
        return EmbeddedFileHelpers.EmbeddedStreamExists($"help.{topic}.txt");
    }

    public static string? GetHelpTopicText(string topic)
    {
        return EmbeddedFileHelpers.ReadEmbeddedStream($"help.{topic}.txt");
    }

    public static void DisplayUsage(string? helpTopic)
    {
        var validTopic = !string.IsNullOrEmpty(helpTopic) && HelpTopicExists(helpTopic);
        var helpContent = validTopic
            ? GetHelpTopicText(helpTopic!)
            : GetHelpTopicText(UsageHelpTopic);

        helpContent ??= $"USAGE: {ProgramInfo.Name} [...]";

        ConsoleHelpers.WriteLine(helpContent.TrimEnd(), overrideQuiet: true);
    }

    public static void DisplayHelpTopic(string topic, bool expandTopics = false)
    {
        topic ??= UsageHelpTopic;

        var helpTopicExists = HelpTopicExists(topic);
        if (!helpTopicExists)
        {
            if (string.IsNullOrEmpty(topic))
            {
                DisplayHelpTopic("help");
                return;
            }

            if (topic == "topics" || topic == "topics expand")
            {
                expandTopics = expandTopics || topic == "topics expand";
                DisplayHelpTopics(expandTopics);
                return;
            }

            if (topic.StartsWith("find"))
            {
                topic = topic.Substring("find".Length).Trim();
                var helpTopics = GetHelpTopics().Where(t => HelpTopicContains(t, topic)).ToList();
                if (helpTopics.Count > 0)
                {
                    DisplayHelpTopics(helpTopics, expandTopics);
                    return;
                }
            }
            
            if (topic.StartsWith("ama ", StringComparison.OrdinalIgnoreCase))
            {
                string question = topic.Substring("ama ".Length).Trim();
                if (!string.IsNullOrEmpty(question))
                {
                    string toolName = ProgramInfo.Name.ToLower();
                    string command = $"cycod --inputs \"/run {toolName} help topics --expand\" \"Please tell me how to: {question}\" \"exit\"";
                    var result = ProcessHelpers.RunProcess(command);
                    
                    // Clean up the output to show only the AI's response
                    string output = result.MergedOutput;
                    
                    // Find the start of the response (after "Please tell me how to: {question}")
                    string startPattern = $"User: Please tell me how to: {question}";
                    int startIndex = output.IndexOf(startPattern);
                    if (startIndex >= 0)
                    {
                        // Move past this line to get to the assistant's response
                        startIndex = output.IndexOf('\n', startIndex);
                        if (startIndex >= 0) startIndex++;
                    }
                    
                    // Find the end of the response (before "User: exit")
                    string endPattern = "User: exit";
                    int endIndex = output.IndexOf(endPattern);
                    
                    // Extract just the assistant's response
                    if (startIndex >= 0 && endIndex > startIndex)
                    {
                        output = output.Substring(startIndex, endIndex - startIndex).Trim();
                        
                        // Remove "Assistant:" prefix if present
                        if (output.StartsWith("Assistant:"))
                        {
                            output = output.Substring("Assistant:".Length).Trim();
                        }
                        // Also check for "A:" prefix
                        else if (output.StartsWith("A:"))
                        {
                            output = output.Substring("A:".Length).Trim();
                        }
                    }
                    
                    ConsoleHelpers.WriteLine(output, overrideQuiet: true);
                    return;
                }
            }

            ConsoleHelpers.WriteLine(
                $"  WARNING: No help topic found for '{topic}'\n\n" +
                "    " + GetHelpTopicText("help")?.Replace("\n", "\n    "),
                overrideQuiet: true
                );
            return;
        }

        var helpContent = GetHelpTopicText(topic)!;
        ConsoleHelpers.WriteLine(helpContent.TrimEnd(), overrideQuiet: true);
    }

    public static void DisplayHelpTopics(bool expandTopics)
    {
        var helpTopics = GetHelpTopics();
        DisplayHelpTopics(helpTopics, expandTopics);
    }

    public static void DisplayHelpTopics(IEnumerable<string> topics, bool expandTopics)
    {
        topics = topics.Select(t => expandTopics
            ? $"## `{ProgramInfo.Name} help {t}`\n\n```\n{GetHelpTopicText(t)}\n```\n"
            : $"  {ProgramInfo.Name} help {t}").ToList();
        ConsoleHelpers.WriteLine(string.Join("\n", topics), overrideQuiet: true);
    }

    private static bool HelpTopicContains(string topic, string searchFor)
    {
        var nameMatches = topic.Contains(searchFor, StringComparison.OrdinalIgnoreCase);
        var contentMatches = GetHelpTopicText(topic)?.Contains(searchFor, StringComparison.OrdinalIgnoreCase) ?? false;
        return nameMatches || contentMatches;
    }

    private const string UsageHelpTopic = "usage";
}

File: src/commonjs/Helpers/HelpHelpers.ts
import { EmbeddedFileHelpers } from './EmbeddedFileHelpers';
import { ProgramInfo } from './ProgramInfo';
import { ConsoleHelpers } from './ConsoleHelpers';

/**
 * Helper class for managing help topics and displaying help content.
 */
export class HelpHelpers {
    private static readonly UsageHelpTopic = "usage";

    /**
     * Gets all available help topics from embedded resources.
     * @returns Array of help topic names
     */
    public static GetHelpTopics(): string[] {
        const allResourceNames = EmbeddedFileHelpers.GetEmbeddedStreamFileNames();

        const helpPrefix = `${ProgramInfo.Name}.help.`;
        const helpTopics = allResourceNames
            .filter(name => name.toLowerCase().startsWith(helpPrefix.toLowerCase()))
            .map(name => name.substring(helpPrefix.length))
            .map(name => name.substring(0, name.length - '.txt'.length))
            .filter((value, index, self) => self.indexOf(value) === index) // Distinct
            .sort((a, b) => {
                // Sort by space count first, then alphabetically
                const aSpaceCount = (a.match(/ /g) || []).length;
                const bSpaceCount = (b.match(/ /g) || []).length;
                const spaceDiff = aSpaceCount - bSpaceCount;
                return spaceDiff !== 0 ? spaceDiff : a.localeCompare(b);
            });

        return helpTopics;
    }

    /**
     * Checks if a help topic exists.
     * @param topic The help topic name
     * @returns True if the topic exists
     */
    public static HelpTopicExists(topic: string): boolean {
        return EmbeddedFileHelpers.EmbeddedStreamExists(`help.${topic}.txt`);
    }

    /**
     * Gets the text content for a help topic.
     * @param topic The help topic name
     * @returns The help topic text, or null if not found
     */
    public static GetHelpTopicText(topic: string): string | null {
        return EmbeddedFileHelpers.ReadEmbeddedStream(`help.${topic}.txt`);
    }

    /**
     * Displays usage information for a specific help topic or general usage.
     * @param helpTopic The help topic to display, or null for general usage
     */
    public static DisplayUsage(helpTopic: string | null): void {
        const validTopic = helpTopic && helpTopic.trim() !== '' && HelpHelpers.HelpTopicExists(helpTopic);
        const helpContent = validTopic
            ? HelpHelpers.GetHelpTopicText(helpTopic)
            : HelpHelpers.GetHelpTopicText(HelpHelpers.UsageHelpTopic);

        const finalContent = helpContent ?? `USAGE: ${ProgramInfo.Name} [...]`;

        ConsoleHelpers.WriteLine(finalContent.trimEnd(), undefined, true);
    }

    /**
     * Displays a specific help topic or handles special topic commands.
     * @param topic The help topic to display
     * @param expandTopics Whether to expand topic listings with full content
     */
    public static DisplayHelpTopic(topic: string, expandTopics: boolean = false): void {
        const actualTopic = topic ?? HelpHelpers.UsageHelpTopic;

        const helpTopicExists = HelpHelpers.HelpTopicExists(actualTopic);
        if (!helpTopicExists) {
            if (!actualTopic || actualTopic.trim() === '') {
                HelpHelpers.DisplayHelpTopic("help");
                return;
            }

            if (actualTopic === "topics" || actualTopic === "topics expand") {
                const shouldExpand = expandTopics || actualTopic === "topics expand";
                HelpHelpers.DisplayHelpTopics(shouldExpand);
                return;
            }

            if (actualTopic.startsWith("find")) {
                const searchTerm = actualTopic.substring("find".length).trim();
                const matchingTopics = HelpHelpers.GetHelpTopics().filter(t => 
                    HelpHelpers.HelpTopicContains(t, searchTerm));
                
                if (matchingTopics.length > 0) {
                    HelpHelpers.DisplayHelpTopicsArray(matchingTopics, expandTopics);
                    return;
                }
            }

            const helpTopicText = HelpHelpers.GetHelpTopicText("help");
            const indentedHelp = helpTopicText?.replace(/\n/g, "\n    ") || "";
            
            ConsoleHelpers.WriteLine(
                `  WARNING: No help topic found for '${actualTopic}'\n\n    ${indentedHelp}`,
                undefined,
                true
            );
            return;
        }

        const helpContent = HelpHelpers.GetHelpTopicText(actualTopic);
        if (helpContent) {
            ConsoleHelpers.WriteLine(helpContent.trimEnd(), undefined, true);
        }
    }

    /**
     * Displays all available help topics.
     * @param expandTopics Whether to expand topics with full content
     */
    public static DisplayHelpTopics(expandTopics: boolean): void {
        const helpTopics = HelpHelpers.GetHelpTopics();
        HelpHelpers.DisplayHelpTopicsArray(helpTopics, expandTopics);
    }

    /**
     * Displays a specific array of help topics.
     * @param topics The topics to display
     * @param expandTopics Whether to expand topics with full content
     */
    public static DisplayHelpTopicsArray(topics: string[], expandTopics: boolean): void {
        const formattedTopics = topics.map(t => expandTopics
            ? `## \`${ProgramInfo.Name} help ${t}\`\n\n\`\`\`\n${HelpHelpers.GetHelpTopicText(t) || ''}\n\`\`\`\n`
            : `  ${ProgramInfo.Name} help ${t}`);
        
        ConsoleHelpers.WriteLine(formattedTopics.join("\n"), undefined, true);
    }

    /**
     * Checks if a help topic contains the specified search term.
     * @param topic The topic name to check
     * @param searchFor The search term
     * @returns True if the topic name or content contains the search term
     */
    private static HelpTopicContains(topic: string, searchFor: string): boolean {
        const nameMatches = topic.toLowerCase().includes(searchFor.toLowerCase());
        const topicText = HelpHelpers.GetHelpTopicText(topic);
        const contentMatches = topicText?.toLowerCase().includes(searchFor.toLowerCase()) ?? false;
        return nameMatches || contentMatches;
    }
}
